### PR TITLE
[New3D] Multi-object picking with context menu

### PR DESCRIPTION
**Pull request by jhurliman**
**User-Facing Changes**

- `3D (Beta)` shows a context menu when clicking on multiple objects

**Description**

Renderer now emits a `renderablesClicked` event, and external code in ThreeDeeRender is responsible for actually selecting a single renderable, either immediately if the list contains one entry or later through a context menu action if there are multiple candidate renderables.

https://user-images.githubusercontent.com/195374/179289637-2c6c72c5-80eb-4585-8e2d-ccf4ba30324e.mov

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/Interactions/Interactions.tsx [hunk 1]
```diff
@@ -54,19 +50,8 @@ const InteractionsBaseComponent = React.memo<Props>(function InteractionsBaseCom
...
-  const { originalMessage } = selectedInteractionData ?? {};
-
-  const isPointCloud = maybeCast<{ type?: number }>(object)?.type === 102;
-  const maybeFullyDecodedObject = React.useMemo(
-    () =>
-      isPointCloud
-        ? decodeAdditionalFields(originalMessage as unknown as PointCloud2)
-        : originalMessage,
-    [isPointCloud, originalMessage],
-  );
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/Interactions/Interactions.tsx:69]: Is the ability to click on individual points in a point cloud (and see the decoded field values) not implemented yet?

File: packages/studio-base/src/panels/ThreeDeeRender/Renderable.ts [hunk 2]
```diff
@@ -31,9 +31,13 @@ export type BaseUserData = {
  * automatically be applied and settings tree errors to be displayed.
  */
 export class Renderable<TUserData extends BaseUserData = BaseUserData> extends THREE.Object3D {
+  /** Identifies this class as inheriting from `Renderable` */
   readonly isRenderable = true;
-  readonly pickable: boolean = true;
+  /** Allow this object to be selected during picking and shown in the Object Details view */
+  readonly picking: boolean = true;
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/Renderable.ts:37]: pickable made more sense as a name than picking IMO

File: packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx [hunk 3]
```diff
@@ -77,11 +78,11 @@ function RendererOverlay(props: {
...
   }, [interactionsTabType, renderer]);
 
-  useRendererEvent("renderableSelected", (renderable) => {
-    setSelectedRenderable(renderable);
-    if (renderable) {
-      setInteractionsTabType(OBJECT_TAB_TYPE);
-    }
+  useRendererEvent("renderablesClicked", (renderables, cursorCoords) => {
+    const rect = props.canvas!.getBoundingClientRect();
+    setClickedPosition({ clientX: rect.left + cursorCoords.x, clientY: rect.top + cursorCoords.y });
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx:159]: are we somewhere converting clientXY into cursorCoords and just converting back?

File: packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx [hunk 4]
```diff
@@ -135,6 +154,7 @@ function RendererOverlay(props: {
           flexDirection: "column",
           alignItems: "flex-end",
           gap: 10,
+          pointerEvents: "none",
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx:244]: Do we not need to re-enable pointerEvents:"auto" on some children to make them clickable?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/MeasurementTool.ts [hunk 5]
```diff
@@ -68,6 +68,10 @@ export class MeasurementTool extends SceneExtension {
   constructor(renderer: Renderer) {
     super("foxglove.MeasurementTool", renderer);
 
+    this.line.userData.picking = false;
+    this.circle1.userData.picking = false;
+    this.circle2.userData.picking = false;
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/MeasurementTool.ts:81]: could you also update the PublishClickTool which I just merged? I had something like this in there already which may need to change.

**Review by jtbandes — APPROVED** [review]

File: packages/studio-base/src/panels/ThreeDeeRender/Interactions/Interactions.tsx (hunk 1 above)
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/Interactions/Interactions.tsx:69]: Right, that's a separate ticket. The pointcloud-specific UI will go away regardless, we need a more generic way of supporting fine-grained picking for custom renderables (this was a customer request).

File: packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx (hunk 3 above)
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx:159]: Possibly, I'm not sure. `cursorCoords` is relative to the canvas, but `clickedPosition` needs to be relative to the browser viewport.

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/MeasurementTool.ts (hunk 5 above)
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/MeasurementTool.ts:81]: I just checked and it looks fine.

```
    this.sphere.visible = false;
    this.sphere.mesh.userData.picking = false;
    this.arrow.visible = false;
    this.arrow.shaftMesh.userData.picking = false;
    this.arrow.headMesh.userData.picking = false;
```

That should do it.

**Pull request merged**